# Chapter3 案例练习——Python数据预处理

In [1]:
import pandas as pd
import numpy as np

In [2]:
# pandas2.2版本以上可以通过calamine读取加速
!pip install python-calamine

Defaulting to user installation because normal site-packages is not writeable


In [3]:
# import warnings
# warnings.filterwarnings('ignore')

## 数据读取

In [4]:
dtype = {
    "InvoiceNo": "string",
    "StockCode": "string",
    "CustomerID": "string",
    "Country": "string",
    "Description": "string",
    "Quantity": "int32",
    "UnitPrice": "float32",
}

path = '../data/Online Retail.xlsx'
sales = pd.read_excel(path, sheet_name='Online Retail', dtype=dtype, engine='calamine')

In [5]:
sales

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[541909 rows x 8 columns]

## 数据属性

In [6]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  string        
 1   StockCode    541909 non-null  string        
 2   Description  540455 non-null  string        
 3   Quantity     541909 non-null  int32         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float32       
 6   CustomerID   406829 non-null  string        
 7   Country      541909 non-null  string        
dtypes: datetime64[ns](1), float32(1), int32(1), string(5)
memory usage: 28.9 MB


In [7]:
sales.dtypes

InvoiceNo      string[python]
StockCode      string[python]
Description    string[python]
Quantity                int32
InvoiceDate    datetime64[ns]
UnitPrice             float32
CustomerID     string[python]
Country        string[python]
dtype: object

### 数据类型转换

In [8]:
# sales.InvoiceNo = sales.InvoiceNo.astype('string')
# sales.StockCode = sales.StockCode.astype('string')
# sales.Description = sales.Description.astype('string')
# sales.CustomerID = sales.CustomerID.astype('string')
# sales.Country = sales.Country.astype('string')
# sales.Quantity = sales.Quantity.astype('int')
# sales.UnitPrice = sales.UnitPrice.astype('float')
# sales.InvoiceDate = pd.to_datetime(sales.InvoiceDate)

In [9]:
sales.dtypes

InvoiceNo      string[python]
StockCode      string[python]
Description    string[python]
Quantity                int32
InvoiceDate    datetime64[ns]
UnitPrice             float32
CustomerID     string[python]
Country        string[python]
dtype: object

## 数据清洗

### 重复值

#### 查找

In [10]:
sales.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Length: 541909, dtype: bool

In [11]:
sales[sales.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

               InvoiceDate  UnitPrice CustomerID         Country  
517    2010-12-01 11:45:00       1.25      17908  United Kingdom  
527    2010-12-01 11:45:00       2.10      17908  United Kingdom  
537    2010-12-01 11:45:00       2.95      17908  United Kingdom  
539    2010-12-01 11:45:00       4.95      17908  United Kingdom  
555    2010-12-01 11:49:00       2.95      17920  United Kingdom  
...                    ...        ...        ...             ...  
541675 2011-12-09 11:34:00       0.39      14446  United Kingdom  
541689 2011-12-09 11:34:00       2.49      14446  United Kingdom  
541692 2011-12-09 11:34:00       1.95      14446  United Kingdom  
541699 2011-12-09 11:34:00       2.10      14446  United Kingdom  
541701 2011-12-09 11:34:00       2.08      14446  United Kingdom  

[5268 rows x 8 columns]

#### 删除

In [12]:
row_before = sales.shape[0]
row_before

541909

In [13]:
sales.drop_duplicates(inplace=True)

In [14]:
row_after = sales.shape[0]
row_after

536641

In [15]:
row_before - row_after

5268

In [16]:
sales

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[536641 rows x 8 columns]

重新设置行标签

In [17]:
sales.reset_index(drop=True, inplace=True)

In [18]:
sales

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536636    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
536637    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
536638    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
536639    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
536640    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
536636 2011-12-09 12:50:00       0.85      12680          France  
536637 2011-12-09 12:50:00       2.10      12680          France  
536638 2011-12-09 12:50:00       4.15      12680          France  
536639 2011-12-09 12:50:00       4.15      12680          France  
536640 2011-12-09 12:50:00       4.95      12680          France  

[536641 rows x 8 columns]

### 缺失值

In [19]:
sales

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536636    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
536637    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
536638    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
536639    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
536640    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
536636 2011-12-09 12:50:00       0.85      12680          France  
536637 2011-12-09 12:50:00       2.10      12680          France  
536638 2011-12-09 12:50:00       4.15      12680          France  
536639 2011-12-09 12:50:00       4.15      12680          France  
536640 2011-12-09 12:50:00       4.95      12680          France  

[536641 rows x 8 columns]

#### 查找缺失值

In [20]:
sales.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135037
Country             0
dtype: int64

In [21]:
sales[sales.isnull().values==True]

InvoiceNo StockCode                      Description  Quantity  \
605       536414     22139                             <NA>        56   
605       536414     22139                             <NA>        56   
1407      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1408      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1409      536544     21786               POLKADOT RAIN HAT          4   
...          ...       ...                              ...       ...   
536275    581498    85099B          JUMBO BAG RED RETROSPOT         5   
536276    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
536277    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
536278    581498     85174                S/4 CACTI CANDLES         1   
536279    581498       DOT                   DOTCOM POSTAGE         1   

               InvoiceDate    UnitPrice CustomerID         Country  
605    2010-12-01 11:52:00     0.000000       <NA>  United Kingdom  
605    2010-12-01 11:52:00     0.000000       <NA>  United Kingdom  
1407   2010-12-01 14:32:00     2.510000       <NA>  United Kingdom  
1408   2010-12-01 14:32:00     2.510000       <NA>  United Kingdom  
1409   2010-12-01 14:32:00     0.850000       <NA>  United Kingdom  
...                    ...          ...        ...             ...  
536275 2011-12-09 10:26:00     4.130000       <NA>  United Kingdom  
536276 2011-12-09 10:26:00     4.130000       <NA>  United Kingdom  
536277 2011-12-09 10:26:00     4.960000       <NA>  United Kingdom  
536278 2011-12-09 10:26:00    10.790000       <NA>  United Kingdom  
536279 2011-12-09 10:26:00  1714.170044       <NA>  United Kingdom  

[136491 rows x 8 columns]

分析：
+ Description有1454缺失值，删除不合适，也比较难填充
+ CustomerID有135037缺失值，删除不合适，也比较难填充，可以统一填充为一个编号代表匿名顾客

#### 处理

先确认下数据类型，免得填充数据类型不匹配

In [22]:
sales.dtypes

InvoiceNo      string[python]
StockCode      string[python]
Description    string[python]
Quantity                int32
InvoiceDate    datetime64[ns]
UnitPrice             float32
CustomerID     string[python]
Country        string[python]
dtype: object

##### CustomerID

用'0'代表匿名顾客，进行填充，先确认下有没有CustomerID为'0'的数据

In [23]:
sales[sales.CustomerID == '0']

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

缺失CustomerID填充'0'，代表匿名顾客

In [24]:
sales.CustomerID = sales.CustomerID.fillna('0')

In [25]:
sales.isnull().sum()

InvoiceNo         0
StockCode         0
Description    1454
Quantity          0
InvoiceDate       0
UnitPrice         0
CustomerID        0
Country           0
dtype: int64

##### Description

Descripition填充为null

In [26]:
sales.Description = sales.Description.fillna('null')

In [27]:
sales.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

### 一致化

### 异常值

In [28]:
sales.describe()

Quantity                    InvoiceDate      UnitPrice
count  536641.000000                         536641  536641.000000
mean        9.620029  2011-07-04 08:57:06.087421952       4.632655
min    -80995.000000            2010-12-01 08:26:00  -11062.059570
25%         1.000000            2011-03-28 10:52:00       1.250000
50%         3.000000            2011-07-19 14:04:00       2.080000
75%        10.000000            2011-10-18 17:05:00       4.130000
max     80995.000000            2011-12-09 12:50:00   38970.000000
std       219.130156                            NaN      97.187050

分析：
+ 异常值存在4种情况
+ Quantity = 0 或 Quantity < 0
+ UnitPrice = 0 或 UnitPrice < 0

#### Quantity == 0

In [29]:
sales[sales.Quantity == 0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

#### Quantity < 0

##### 查找

In [30]:
sales[sales.Quantity < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
535188   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
536280   C581499         M                            Manual        -1   
536447   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
536448   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
536449   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate   UnitPrice CustomerID         Country  
141    2010-12-01 09:41:00   27.500000      14527  United Kingdom  
154    2010-12-01 09:49:00    4.650000      15311  United Kingdom  
235    2010-12-01 10:24:00    1.650000      17548  United Kingdom  
236    2010-12-01 10:24:00    0.290000      17548  United Kingdom  
237    2010-12-01 10:24:00    0.290000      17548  United Kingdom  
...                    ...         ...        ...             ...  
535188 2011-12-09 09:57:00    0.830000      14397  United Kingdom  
536280 2011-12-09 10:28:00  224.690002      15498  United Kingdom  
536447 2011-12-09 11:57:00   10.950000      15311  United Kingdom  
536448 2011-12-09 11:58:00    1.250000      17315  United Kingdom  
536449 2011-12-09 11:58:00    1.250000      17315  United Kingdom  

[10587 rows x 8 columns]

分析：
+ 一共有10587条记录
+ 可能由于订单编号中有C，是退货订单

In [31]:
query_cancel = sales.InvoiceNo.str.contains('C')
query_cancel

0         False
1         False
2         False
3         False
4         False
          ...  
536636    False
536637    False
536638    False
536639    False
536640    False
Name: InvoiceNo, Length: 536641, dtype: boolean

In [32]:
sales[query_cancel]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
535188   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
536280   C581499         M                            Manual        -1   
536447   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
536448   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
536449   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate   UnitPrice CustomerID         Country  
141    2010-12-01 09:41:00   27.500000      14527  United Kingdom  
154    2010-12-01 09:49:00    4.650000      15311  United Kingdom  
235    2010-12-01 10:24:00    1.650000      17548  United Kingdom  
236    2010-12-01 10:24:00    0.290000      17548  United Kingdom  
237    2010-12-01 10:24:00    0.290000      17548  United Kingdom  
...                    ...         ...        ...             ...  
535188 2011-12-09 09:57:00    0.830000      14397  United Kingdom  
536280 2011-12-09 10:28:00  224.690002      15498  United Kingdom  
536447 2011-12-09 11:57:00   10.950000      15311  United Kingdom  
536448 2011-12-09 11:58:00    1.250000      17315  United Kingdom  
536449 2011-12-09 11:58:00    1.250000      17315  United Kingdom  

[9251 rows x 8 columns]

分析：
+ 只有9251条是因为退货订单造成的
+ 剩下1336条，是什么呢？

##### 处理

把这部分退货订单分出来，后面可以单独分析退货原因

In [33]:
sales_cancel = sales[query_cancel].copy() # 注意此处有copy，从原始数据中copy出来一份，以免修改后覆盖原数据
sales_cancel

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
535188   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
536280   C581499         M                            Manual        -1   
536447   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
536448   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
536449   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate   UnitPrice CustomerID         Country  
141    2010-12-01 09:41:00   27.500000      14527  United Kingdom  
154    2010-12-01 09:49:00    4.650000      15311  United Kingdom  
235    2010-12-01 10:24:00    1.650000      17548  United Kingdom  
236    2010-12-01 10:24:00    0.290000      17548  United Kingdom  
237    2010-12-01 10:24:00    0.290000      17548  United Kingdom  
...                    ...         ...        ...             ...  
535188 2011-12-09 09:57:00    0.830000      14397  United Kingdom  
536280 2011-12-09 10:28:00  224.690002      15498  United Kingdom  
536447 2011-12-09 11:57:00   10.950000      15311  United Kingdom  
536448 2011-12-09 11:58:00    1.250000      17315  United Kingdom  
536449 2011-12-09 11:58:00    1.250000      17315  United Kingdom  

[9251 rows x 8 columns]

从当前订单中分离出成功订单，保存为sales_success

In [34]:
sales_success = sales[-query_cancel].copy() # 注意此处有copy，从原始数据中copy出来一份，代表销售成功数据，以免修改后覆盖原数据
sales_success

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536636    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
536637    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
536638    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
536639    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
536640    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
536636 2011-12-09 12:50:00       0.85      12680          France  
536637 2011-12-09 12:50:00       2.10      12680          France  
536638 2011-12-09 12:50:00       4.15      12680          France  
536639 2011-12-09 12:50:00       4.15      12680          France  
536640 2011-12-09 12:50:00       4.95      12680          France  

[527390 rows x 8 columns]

#### UnitPrice == 0

##### 查找

In [35]:
sales_success[sales_success.UnitPrice == 0]

InvoiceNo StockCode                   Description  Quantity  \
605       536414     22139                          null        56   
1934      536545     21134                          null         1   
1935      536546     22145                          null         1   
1936      536547     37509                          null         1   
1951      536549    85226A                          null         1   
...          ...       ...                           ...       ...   
531753    581234     72817                          null        27   
533259    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
533260    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
533309    581408     85175                          null        20   
533665    581422     23169                       smashed      -235   

               InvoiceDate  UnitPrice CustomerID         Country  
605    2010-12-01 11:52:00        0.0          0  United Kingdom  
1934   2010-12-01 14:32:00        0.0          0  United Kingdom  
1935   2010-12-01 14:33:00        0.0          0  United Kingdom  
1936   2010-12-01 14:33:00        0.0          0  United Kingdom  
1951   2010-12-01 14:34:00        0.0          0  United Kingdom  
...                    ...        ...        ...             ...  
531753 2011-12-08 10:33:00        0.0          0  United Kingdom  
533259 2011-12-08 13:58:00        0.0          0  United Kingdom  
533260 2011-12-08 13:58:00        0.0          0  United Kingdom  
533309 2011-12-08 14:06:00        0.0          0  United Kingdom  
533665 2011-12-08 15:24:00        0.0          0  United Kingdom  

[2510 rows x 8 columns]

分析：
+ 这部分可能为免费订单
+ 这部分数据CustomerID很多为0，且Description为null，非常蹊跷
+ 免费订单和Quantity < 0会不会有关？

In [36]:
sales_success[(sales_success.Quantity < 0) & (sales_success.UnitPrice == 0)]

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
2362      536589     21777        null       -10 2010-12-01 16:50:00   
4286      536764    84952C        null       -38 2010-12-02 14:42:00   
7086      536996     22712        null       -20 2010-12-03 15:30:00   
7087      536997     22028        null       -20 2010-12-03 15:30:00   
7088      536998     85067        null        -6 2010-12-03 15:30:00   
...          ...       ...         ...       ...                 ...   
530105    581210     23395       check       -26 2011-12-07 18:36:00   
530107    581212     22578        lost     -1050 2011-12-07 18:38:00   
530108    581213     22576       check       -30 2011-12-07 18:38:00   
531680    581226     23090     missing      -338 2011-12-08 09:56:00   
533665    581422     23169     smashed      -235 2011-12-08 15:24:00   

        UnitPrice CustomerID         Country  
2362          0.0          0  United Kingdom  
4286          0.0          0  United Kingdom  
7086          0.0          0  United Kingdom  
7087          0.0          0  United Kingdom  
7088          0.0          0  United Kingdom  
...           ...        ...             ...  
530105        0.0          0  United Kingdom  
530107        0.0          0  United Kingdom  
530108        0.0          0  United Kingdom  
531680        0.0          0  United Kingdom  
533665        0.0          0  United Kingdom  

[1336 rows x 8 columns]

分析：
+ UnitPrice = 0且Quantity < 0的数据刚好为1336，说明前面残留都Quantity < 0可能是单价为0的数据造成
+ 看Description中大部分都是物品丢失或损坏

##### 处理

这部分UnitPrice为0的记录并不会影响最后营收，可以单独分出来后面分析原因

In [37]:
query_free = sales_success.UnitPrice == 0

In [38]:
sales_free = sales_success[query_free].copy() #注意，此处copy备份，从原始数据中copy出来一份，以免修改后覆盖原数据
sales_free

InvoiceNo StockCode                   Description  Quantity  \
605       536414     22139                          null        56   
1934      536545     21134                          null         1   
1935      536546     22145                          null         1   
1936      536547     37509                          null         1   
1951      536549    85226A                          null         1   
...          ...       ...                           ...       ...   
531753    581234     72817                          null        27   
533259    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
533260    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
533309    581408     85175                          null        20   
533665    581422     23169                       smashed      -235   

               InvoiceDate  UnitPrice CustomerID         Country  
605    2010-12-01 11:52:00        0.0          0  United Kingdom  
1934   2010-12-01 14:32:00        0.0          0  United Kingdom  
1935   2010-12-01 14:33:00        0.0          0  United Kingdom  
1936   2010-12-01 14:33:00        0.0          0  United Kingdom  
1951   2010-12-01 14:34:00        0.0          0  United Kingdom  
...                    ...        ...        ...             ...  
531753 2011-12-08 10:33:00        0.0          0  United Kingdom  
533259 2011-12-08 13:58:00        0.0          0  United Kingdom  
533260 2011-12-08 13:58:00        0.0          0  United Kingdom  
533309 2011-12-08 14:06:00        0.0          0  United Kingdom  
533665 2011-12-08 15:24:00        0.0          0  United Kingdom  

[2510 rows x 8 columns]

In [39]:
sales_success = sales_success[-query_free].copy()
sales_success

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536636    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
536637    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
536638    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
536639    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
536640    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
536636 2011-12-09 12:50:00       0.85      12680          France  
536637 2011-12-09 12:50:00       2.10      12680          France  
536638 2011-12-09 12:50:00       4.15      12680          France  
536639 2011-12-09 12:50:00       4.15      12680          France  
536640 2011-12-09 12:50:00       4.95      12680          France  

[524880 rows x 8 columns]

#### UnitPrice < 0 

##### 查找

In [40]:
sales_success[sales_success.UnitPrice < 0]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
297646   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
297647   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

          UnitPrice CustomerID         Country  
297646 -11062.05957          0  United Kingdom  
297647 -11062.05957          0  United Kingdom

分析：
+ 存在两条调整坏账数据

##### 处理

这部分UnitPrice < 0的数据可以单独分出来后面分析原因

In [41]:
query_debt = sales_success.UnitPrice < 0

In [42]:
sales_debt = sales_success[query_debt].copy() #注意此处copy，从原始数据中copy出来一份，以免修改后覆盖原数据
sales_debt

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
297646   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
297647   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

          UnitPrice CustomerID         Country  
297646 -11062.05957          0  United Kingdom  
297647 -11062.05957          0  United Kingdom

In [43]:
sales_success = sales_success[-query_debt].copy()
sales_success

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536636    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
536637    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
536638    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
536639    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
536640    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
536636 2011-12-09 12:50:00       0.85      12680          France  
536637 2011-12-09 12:50:00       2.10      12680          France  
536638 2011-12-09 12:50:00       4.15      12680          France  
536639 2011-12-09 12:50:00       4.15      12680          France  
536640 2011-12-09 12:50:00       4.95      12680          France  

[524878 rows x 8 columns]

#### 再次确认

In [44]:
sales_success.describe()

Quantity                    InvoiceDate      UnitPrice
count  524878.000000                         524878  524878.000000
mean       10.616600  2011-07-04 15:30:16.317049088       3.922573
min         1.000000            2010-12-01 08:26:00       0.001000
25%         1.000000            2011-03-28 12:13:00       1.250000
50%         4.000000            2011-07-20 11:22:00       2.080000
75%        11.000000            2011-10-19 11:41:00       4.130000
max     80995.000000            2011-12-09 12:50:00   13541.330078
std       156.280031                            NaN      36.043133

In [45]:
sales_success.reset_index(drop=True, inplace=True)

In [46]:
sales_success

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
524873    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
524874    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
524875    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
524876    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
524877    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
524873 2011-12-09 12:50:00       0.85      12680          France  
524874 2011-12-09 12:50:00       2.10      12680          France  
524875 2011-12-09 12:50:00       4.15      12680          France  
524876 2011-12-09 12:50:00       4.15      12680          France  
524877 2011-12-09 12:50:00       4.95      12680          France  

[524878 rows x 8 columns]

### 保存数据

In [47]:
path = '../data/Online Retail Clean.xlsx'
sales_success.to_excel(path, sheet_name='Online Retail Clean', index=False)

## 特征转换

### 货币单位转换

方法一：可以采用标量相乘

In [48]:
exchange_rate = 9.06
sales_success['UnitPrice_RMB'] = sales_success.UnitPrice * exchange_rate

In [49]:
# 未来更推荐这种
sales_success.loc[:, 'UnitPrice_RMB'] = sales_success.UnitPrice * exchange_rate

In [50]:
sales_success.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  UnitPrice_RMB  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom      23.103001  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713402  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom      24.915001  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713402  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713402

方法二：可以用自定义函数

In [51]:
def pound_to_yuan(x, exchange_rate=9.06):
    return x * exchange_rate

In [52]:
sales_success['UnitPrice_RMB'] = sales_success.UnitPrice.map(pound_to_yuan)

In [53]:
# 未来更推荐这种
sales_success.loc[:, 'UnitPrice_RMB'] = sales_success.UnitPrice.map(pound_to_yuan)

In [54]:
sales_success.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  UnitPrice_RMB  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom      23.103000  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713401  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom      24.915000  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713401  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713401

## 数据构造

In [55]:
sales_success['SumPrice']  = sales_success.UnitPrice_RMB * sales_success.Quantity

In [56]:
# 未来更推荐这种
sales_success.loc[:, 'SumPrice']  = sales_success.UnitPrice_RMB * sales_success.Quantity

In [57]:
sales_success.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  UnitPrice_RMB  \
0 2010-12-01 08:26:00       2.55      17850  United Kingdom      23.103000   
1 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713401   
2 2010-12-01 08:26:00       2.75      17850  United Kingdom      24.915000   
3 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713401   
4 2010-12-01 08:26:00       3.39      17850  United Kingdom      30.713401   

     SumPrice  
0  138.617997  
1  184.280406  
2  199.320000  
3  184.280406  
4  184.280406

## GMV计算

In [58]:
sales_success.SumPrice.sum()

96417523.37038195

## 保存数据

In [59]:
path = '../data/Online Retail Preprocessing.xlsx'
sales_success.to_excel(path, sheet_name='Online Retail Preprocessing', index=False)

# 课堂练习